text/0000-code-literals.md

```diff
@@ -0,0 +1,517 @@
+    if the code is ever logged (such as might happen when logging
+    SQL query executions).
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+A modifier `h` (for
+[Here document](https://en.wikipedia.org/wiki/Here_document))
+may be added to a string literal prefix to change how the
+string is interpreted by the compiler. The effect of the `h`
+modifier causes all indentation to be relative to the
+closing quote:
+
+```rust
+    let code = h"
+        This is a code string literal.
+
+        I can use escape sequences like \n since the `h`
+        prefix was added to a normal string literal
+
+            Indentation is preserved *relative* to the indentation level
+            of the terminating quote.
+
+    It is an error for a line to have negative indentation (ie. be
+    indented less than the final quote) unless
+    the line is empty.
+        ";
+```
+
+`rustfmt` will automatically adjust the indentation of the code string
+literal as a whole to match the surrounding context, but will never
+change the relative indentation within such a literal.
+
+The `h` modifier will often be combined with raw string literals to
+embed sections of code such as SQL:
+
+```rust
+    let code = hr#"
+        This is also a code string literal
+
+        I can use special characters like "" and \ freely.
+
+            Indentation is still *relative* to the indentation level
+            of the terminating quote.
+        "#;
+```
+
+For completeness, the `h` modifier may also be combined with byte
+and raw byte string literals, eg. `hb"` and `hbr#"`.
+
+Anything directly after the opening quote is not considered
+part of the string literal. It may be used as a language hint or
+processed by macros (similar to the treatment of doc comments).
+
+```rust
+let sql = hr#"sql
+    SELECT * FROM table;
+    "#;
+```
+
+When the `h` modifier is used with a raw string literal, the same
+rules as usual apply, where the number of `#` characters can be
+increased if the sequence `"#` needs to appear inside the string.
+
+In order to suppress the final newline, the literal may instead be
+closed with `-" ` or `-"#` depending on the opening quote, eg.
+
+```rust
+let code = hr#"
+    Text with no final newline
+    -"#;
+```
+
+Aside from this `-` modifier, only whitespace may appear on the final
+line prior to the closing quote.
+
+Together, these rules ensure that every possible string can be represented
+in a single canonical way, while allowing the indentation of the string
+as a whole to be changed freely.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+An `h` modifier may be added to the prefix of the following string
+literal types:
+
+- String literals `h"`
+- Raw string literals `hr#"`
+- Byte string literals `hb"`
+- Raw byte string literals `hbr#"`
+
+The `h` modifier will appear before all characters in the prefix.
```

Comment:
There is no persuasive and particular reason offered to have this precede all other characters in the prefix. It would be preferable to assume that we are going to explore accepting a non-canonical ordering.

Reply:
Experimentally, `br"<content>"` compiles, but `rb"<content>"` does not compile. This implies that we are already particular about the order of string prefixes, and so I wrote this RFC with consistency in mind. I don't particularly care about what order is "canonical" but this rules was easy to define and seemed reasonably intuitive. If you have a strong reason to prefer a different order I'd love to heear it.